src/map.rs

```diff
@@ -532,6 +532,198 @@ where
+                                // NOTE: now_garbage == current_value
+
+                                // safety: need to guarantee that now_garbage is no longer
+                                // reachable. more specifically, no thread that executes _after_
+                                // this line can ever get a reference to now_garbage.
+                                //
+                                // here are the possible cases:
+                                //
+                                //  - another thread already has a reference to now_garbage.
+                                //    they must have read it before the call to swap.
+                                //    because of this, that thread must be pinned to an epoch <=
+                                //    the epoch of our guard. since the garbage is placed in our
+                                //    epoch, it won't be freed until the _next_ epoch, at which
+                                //    point, that thread must have dropped its guard, and with it,
+                                //    any reference to the value.
+                                //  - another thread is about to get a reference to this value.
+                                //    they execute _after_ the swap, and therefore do _not_ get a
+                                //    reference to now_garbage (they get value instead). there are
+                                //    no other ways to get to a value except through its Node's
+                                //    `value` field (which is what we swapped), so freeing
+                                //    now_garbage is fine.
+                                unsafe { guard.defer_destroy(now_garbage) };
+                            } else {
+                                delta = -1;
+                                // remove the BinEntry containing the removed key value pair from the bucket
+                                if !pred.is_null() {
+                                    // either by changing the pointer of the previous BinEntry, if present
+                                    // safety: see remove
+                                    unsafe { pred.deref() }
+                                        .as_node()
+                                        .unwrap()
+                                        .next
+                                        .store(next, Ordering::SeqCst);
+                                } else {
+                                    // or by setting the next node as the first BinEntry if there is no previous entry
+                                    t.store_bin(bini, next);
+                                }
+
+                                // in either case, mark the BinEntry as garbage, since it was just removed
+                                // safety: see remove
+                                unsafe { guard.defer_destroy(p) };
```

Comment:
In this case you _also_ need to `defer_destroy` the `.value` _inside_ `p`. Notice how `remove` does that [here](https://github.com/jonhoo/flurry/blob/a5e7cea119e6fd29bc1ab5748c35c0987f81df90/src/map.rs#L1079).

Reply:
Is this resolved by https://github.com/jonhoo/flurry/commit/729c82f571141f65344f4a2e89d5a64efc7c419e?

Reply:
Yup!

Reply:
Yes, although it'd be good to also add a short note about why we also need to free the value.